enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,246 @@
+authors:
+  - "@dhansen"
+reviewers:
+  - "@ironcladlou"
+  - "@Miciah"
+  - "@frobware"
+approvers:
+  - "@knobunc"
+  - "@ironcladlou"
+creation-date: 2019-09-27
+last-updated: 2019-09-27
+status: implementable
+see-also: 
+replaces:
+superseded-by:
+---
+
+# Configurable DNS Forwarding
+
+This proposal allows cluster operators the ability to configure DNS name query forwarding.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+1. Is `operator.openshift.io` the best API group for the proposed `Upstream` type?
+2. How are API changes handled for cluster downgrades?
+3. Should IPv6 references for `Nameservers` be removed until OpenShift adds IPv6 support?
+
+## Summary
+
+Currently, the [cluster-dns-operator](https://github.com/openshift/cluster-dns-operator) manages DNS forwarding with a
+static ConfigMap:
+
+```bash
+$ oc get configmap/dns-default -n openshift-dns -o yaml
+apiVersion: v1
+data:
+  Corefile: |
+    .:5353 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+            pods insecure
+            upstream
+            fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        forward . /etc/resolv.conf
+        cache 30
+        reload
+    }
+kind: ConfigMap
+metadata:
+  labels:
+    dns.operator.openshift.io/owning-dns: default
+  name: dns-default
+  namespace: openshift-dns
+  <SNIP>
+```
+
+This proposal will generate this ConfigMap with a `forward` section based on user-provided values of type `Upstream`.
+
+## Motivation
+
+DNS name resolution for pods is currently provided by an instance of [CoreDNS](https://coredns.io) that runs on each
+node in the cluster. If a DNS query is for a hostname within the cluster domain suffix (i.e. `cluster.local`), CoreDNS
+resolves the name. Otherwise, CoreDNS proxies DNS messages to an upstream resolver identified by `/etc/resolv.conf` on
+the corresponding node. Although this provides a consistent and reliable approach for name resolution, it restricts how
+cluster operators manage DNS name queries.
+
+### Goals
+
+1. A well defined API for DNS forwarding configuration that is implementation agnostic.
+2. A minimal API surface that can be expanded to support future DNS forwarding use cases.
+
+### Non-Goals
+
+1. Support every possible DNS forwarding use case.
+2. Configure or manage external DNS providers.
+
+## Proposal
+
+The proposed `Upstream` type defines a schema for associating one or more `Nameservers` to a DNS subdomain identified by
+`Domain`. `Nameservers` are responsible for resolving name queries for `Domain`:
+```go
+type Upstream struct {
+	Domain string `json:"domain"`
+	Nameservers []string `json:"nameservers"`
```

Comment:
Do we consider the length of this unbounded? Should we warn if we exceed 3 entries? Why 3?

```
The different configuration options are:

       nameserver Name server IP address
              Internet address of a  name  server  that  the  resolver  should
              query,  either  an  IPv4  address  (in dot notation), or an IPv6
              address in colon (and possibly dot) notation as  per  RFC  2373.
              Up  to  MAXNS  (currently 3, see <resolv.h>) name servers may be
              listed, one per keyword.  If there  are  multiple  servers,  the
              resolver  library queries them in the order listed.  If no name‐
              server entries are present, the  default  is  to  use  the  name
              server  on  the  local machine.  (The algorithm used is to try a
              name server, and if the query times out, try the next, until out
              of name servers, then repeat trying all the name servers until a
              maximum number of retries are made.)
```

(from `man resolv.conf`)

Reply:
Isn't this unrelated?  The resolv.conf governs how the pod resolver works.  Here we are configuring the dns server we run that is the first resolver the pod resolver would try.  Or did I misunderstand the proposal?

Reply:
Related question over in https://github.com/openshift/enhancements/pull/53#discussion_r329703046 re: whether users should expect resolv.conf semantics here

Reply:
> Do we consider the length of this unbounded? Should we warn if we exceed 3 entries? 

@frobware based on https://github.com/openshift/enhancements/pull/53#discussion_r329701039, I'm updating the enhancement to state https://github.com/openshift/cluster-dns-operator/blob/master/assets/dns/configmap.yaml#L15-L17 will get prepended to `Upstreams`.